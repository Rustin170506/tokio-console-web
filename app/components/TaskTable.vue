<template>
    <UTable
        :columns="columns"
        :rows="tasks"
        :sort="{ column: 'total', direction: 'desc' }"
        :loading="pending"
        @select="select"
    >
        <template #loading-state>
            <div class="flex justify-center items-center h-full">
                <TokioSvg
                    class="w-8 h-auto mr-2 fill-black dark:fill-white my-8 animate-spin"
                />
            </div>
        </template>
        <template #fields-data="{ row }">
            <div v-if="row.fields && row.fields.length > 0" class="mb-1">
                <p>
                    <span class="text-blue-600 dark:text-blue-400">{{
                        row.fields[0].name
                    }}</span>
                    <span class="text-gray-500 dark:text-gray-400">=</span>
                    <span class="text-green-600 dark:text-green-400">{{
                        row.fields[0].value.value
                    }}</span>
                </p>
            </div>
        </template>
        <template #state-data="{ row }">
            <UIcon :name="row.state" dynamic />
        </template>
        <template #total-data="{ row }">
            <p :class="`${row.total.class} w-10`">
                {{ row.total.toString() }}
            </p>
        </template>
        <template #busy-data="{ row }">
            <p :class="`${row.busy.class} w-10`">
                {{ row.busy.toString() }}
            </p>
        </template>
        <template #sched-data="{ row }">
            <p :class="`${row.sched.class} w-10`">
                {{ row.sched.toString() }}
            </p>
        </template>
        <template #idle-data="{ row }">
            <p :class="`${row.idle.class} w-10`">
                {{ row.idle.toString() }}
            </p>
        </template>
    </UTable>
</template>

<script setup lang="ts">
import { toTaskTableItem, type TaskTableItem } from "~/types/taskTableItem";

const columns = [
    {
        key: "idString",
        label: "ID",
        sortable: true,
    },
    {
        key: "name",
        label: "Name",
        sortable: true,
    },
    {
        key: "state",
        label: "State",
        sortable: true,
    },
    {
        key: "total",
        label: "Total",
        sortable: true,
    },
    {
        key: "busy",
        label: "Busy",
        sortable: true,
    },
    {
        key: "sched",
        label: "Scheduled",
        sortable: true,
    },
    {
        key: "idle",
        label: "Idle",
        sortable: true,
    },
    {
        key: "pools",
        label: "Pools",
        sortable: true,
    },
    {
        key: "kind",
        label: "Kind",
        sortable: true,
    },
    {
        key: "location",
        label: "Location",
    },
    {
        key: "fields",
        label: "Fields",
    },
];

const router = useRouter();

// Select a task and navigate to its details page.
const select = (row: TaskTableItem) => {
    router.push(`/tasks/${row.id}`);
};

// Fetch tasks and convert them to a table item.
const { pending, tasksData, lastUpdatedAt } = useTasks();
const tasks = computed(() => {
    const lastUpdated = lastUpdatedAt.value;
    if (!lastUpdated) {
        return [];
    }
    return Array.from(tasksData.value.values()).reduce((acc, task) => {
        const taskTableItem = toTaskTableItem(task, lastUpdated);
        if (taskTableItem) {
            acc.push(taskTableItem);
        }
        return acc;
    }, [] as TaskTableItem[]);
});
</script>
