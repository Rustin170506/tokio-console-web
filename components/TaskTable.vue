<template>
    <UTable
        :columns="columns"
        :rows="tasks"
        :sort="{ column: 'total', direction: 'desc' }"
        :loading="pending"
        @select="select"
    >
        <template #loading-state>
            <div class="flex justify-center items-center h-full">
                <TokioSvg
                    class="w-8 h-auto mr-2 fill-black dark:fill-white my-8 animate-spin"
                />
            </div>
        </template>
        <template #fields-data="{ row }">
            <div v-for="(item, index) in row.fields" :key="index" class="mb-1">
                <p>
                    <span class="text-blue-600 dark:text-blue-400">{{
                        item.name
                    }}</span>
                    <span class="text-gray-500 dark:text-gray-400">=</span>
                    <span class="text-green-600 dark:text-green-400">{{
                        item.value
                    }}</span>
                </p>
            </div>
        </template>
        <template #state-data="{ row }">
            <UIcon :name="row.state" />
        </template>
        <template #total-data="{ row }">
            <p :class="`${row.total.class} w-10`">
                {{ row.total.toString() }}
            </p>
        </template>
        <template #busy-data="{ row }">
            <p :class="`${row.busy.class} w-10`">
                {{ row.busy.toString() }}
            </p>
        </template>
        <template #sched-data="{ row }">
            <p :class="`${row.sched.class} w-10`">
                {{ row.sched.toString() }}
            </p>
        </template>
        <template #idle-data="{ row }">
            <p :class="`${row.idle.class} w-10`">
                {{ row.idle.toString() }}
            </p>
        </template>
    </UTable>
</template>

<script setup lang="ts">
import { toTaskTableItem, type TaskTableItem } from "~/types/taskTableItem";

const columns = [
    {
        key: "idString",
        label: "ID",
        sortable: true,
    },
    {
        key: "name",
        label: "Name",
        sortable: true,
    },
    {
        key: "state",
        label: "State",
        sortable: true,
    },
    {
        key: "total",
        label: "Total",
        sortable: true,
    },
    {
        key: "busy",
        label: "Busy",
        sortable: true,
    },
    {
        key: "sched",
        label: "Scheduled",
        sortable: true,
    },
    {
        key: "idle",
        label: "Idle",
        sortable: true,
    },
    {
        key: "pools",
        label: "Pools",
        sortable: true,
    },
    {
        key: "kind",
        label: "Kind",
        sortable: true,
    },
    {
        key: "location",
        label: "Location",
    },
    {
        key: "fields",
        label: "Fields",
    },
];

const router = useRouter();

const select = (row: TaskTableItem) => {
    router.push(`/tasks/${row.id}`);
};

const { pending, tasksData, lastUpdatedAt } = useTasks();
const tasks = computed(() => {
    // Map to array.
    const tasks = Array.from(tasksData.value.values());
    const taskList: Array<TaskTableItem | undefined> = tasks.map((task) => {
        if (lastUpdatedAt.value === undefined) {
            return undefined;
        }
        return toTaskTableItem(task, lastUpdatedAt.value);
    });
    return taskList.filter((task) => task !== undefined) as TaskTableItem[];
});
</script>
