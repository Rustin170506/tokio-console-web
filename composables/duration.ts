export class Duration {
    seconds: bigint;
    nanos: number;

    constructor(seconds: bigint, nanos: number) {
        this.seconds = seconds;
        this.nanos = nanos;
    }

    asDays(): number {
        return Number(this.seconds) / 86400;
    }

    asHours(): number {
        return Number(this.seconds) / 3600;
    }

    asMinutes(): number {
        return Number(this.seconds) / 60;
    }

    asSeconds(): number {
        return Number(this.seconds);
    }

    asMilliseconds(): number {
        return Number(this.seconds) * 1000 + this.nanos / 1e6;
    }

    asMicroseconds(): number {
        return Number(this.seconds) * 1e6 + this.nanos / 1000;
    }

    subtract(other: Duration): Duration {
        const result = new Duration(
            this.seconds - other.seconds,
            this.nanos - other.nanos,
        );
        return result;
    }

    add(other: Duration): Duration {
        const result = new Duration(
            this.seconds + other.seconds,
            this.nanos + other.nanos,
        );
        return result;
    }

    toString(): string {
        const totalSeconds = this.asSeconds();
        const units = [
            { value: Math.floor(this.asDays()), unit: "d" },
            { value: Math.floor(this.asHours()) % 24, unit: "h" },
            { value: Math.floor(this.asMinutes()) % 60, unit: "m" },
            { value: Math.floor(totalSeconds) % 60, unit: "s" },
            { value: Math.floor(this.asMilliseconds()) % 1000, unit: "ms" },
            { value: Math.floor(this.asMicroseconds()) % 1000, unit: "Âµs" },
            { value: this.nanos, unit: "ns" },
        ];

        let str = "";
        let partsIncluded = 0;

        if (totalSeconds >= 60) {
            // For durations longer than 60 seconds, include up to two non-zero units
            for (const { value, unit } of units) {
                if (value > 0) {
                    str += `${value}${unit} `;
                    partsIncluded++;
                    if (partsIncluded === 2) break;
                }
            }
        } else {
            // For durations shorter than 60 seconds, display the most significant non-zero unit
            for (const { value, unit } of units) {
                if (value > 0) {
                    str = `${value}${unit}`;
                    break;
                }
            }
        }

        return str || `${this.nanos}ns`;
    }
}
