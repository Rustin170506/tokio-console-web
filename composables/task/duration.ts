// The Duration class represents a duration of time.
export class Duration {
    // The number of seconds in the duration.
    seconds: bigint;
    // The number of nanoseconds in the duration.
    nanos: number;

    // The constructor takes the number of seconds and nanoseconds.
    constructor(seconds: bigint, nanos: number) {
        this.seconds = seconds;
        this.nanos = nanos;
        // Normalize the duration so that nanos is less than 1e9 and non-negative.
        this.normalize();
    }

    // Normalize ensures that nanos is less than 1e9 and non-negative.
    normalize(): Duration {
        let adjustment = 0;
        // If nanos is 1e9 or more, we need to adjust seconds and nanos.
        if (this.nanos >= 1e9) {
            adjustment = Math.floor(this.nanos / 1e9);
            this.seconds += BigInt(adjustment);
            this.nanos -= adjustment * 1e9;
        } else if (this.nanos < 0) {
            // If nanos is negative, we need to adjust seconds and nanos.
            adjustment = Math.floor(-this.nanos / 1e9) + 1;
            this.seconds -= BigInt(adjustment);
            this.nanos += adjustment * 1e9;
        }
        return this;
    }

    asDays(): number {
        return Number(this.seconds) / 86400;
    }

    asHours(): number {
        return Number(this.seconds) / 3600;
    }

    asMinutes(): number {
        return Number(this.seconds) / 60;
    }

    asSeconds(): number {
        return Number(this.seconds);
    }

    asMilliseconds(): number {
        return Number(this.seconds) * 1000 + this.nanos / 1e6;
    }

    asMicroseconds(): number {
        return Number(this.seconds) * 1e6 + this.nanos / 1000;
    }

    // subtract returns a new Duration that represents the difference between this duration and another.
    subtract(other: Duration): Duration {
        const seconds = this.seconds - other.seconds;
        const nanos = this.nanos - other.nanos;
        const result = new Duration(seconds, nanos);
        return result.normalize();
    }

    // add returns a new Duration that represents the sum of this duration and another.
    add(other: Duration): Duration {
        const result = new Duration(
            this.seconds + other.seconds,
            this.nanos + other.nanos,
        );
        return result.normalize();
    }

    // greaterThan checks if this duration is greater than another.
    greaterThan(other: Duration): boolean {
        return (
            this.seconds > other.seconds ||
            (this.seconds === other.seconds && this.nanos > other.nanos)
        );
    }

    toString(): string {
        const totalSeconds = this.asSeconds();
        const units = [
            { value: Math.floor(this.asDays()), unit: "d" },
            { value: Math.floor(this.asHours()) % 24, unit: "h" },
            { value: Math.floor(this.asMinutes()) % 60, unit: "m" },
            { value: Math.floor(totalSeconds) % 60, unit: "s" },
            { value: Math.floor(this.asMilliseconds()) % 1000, unit: "ms" },
            { value: Math.floor(this.asMicroseconds()) % 1000, unit: "Âµs" },
            { value: this.nanos, unit: "ns" },
        ];

        let str = "";
        let partsIncluded = 0;

        if (totalSeconds >= 60) {
            // For durations longer than 60 seconds, include up to two non-zero units
            for (const { value, unit } of units) {
                if (value > 0) {
                    str += `${value}${unit} `;
                    partsIncluded++;
                    if (partsIncluded === 2) break;
                }
            }
        } else {
            // For durations shorter than 60 seconds, display the most significant non-zero unit
            for (const { value, unit } of units) {
                if (value > 0) {
                    str = `${value}${unit}`;
                    break;
                }
            }
        }

        return str || `${this.nanos}ns`;
    }

    // valueOf returns the duration as a number of microseconds.
    // Used for sorting.
    valueOf(): number {
        return this.asMicroseconds();
    }

    // now returns the current time as a Duration from the Unix epoch.
    // TODO: use more precise time source. For example, performance.now() in browsers.
    static now(): Duration {
        return new Duration(BigInt(Math.floor(Date.now() / 1000)), 0);
    }

    static fromNano(nanos: bigint): Duration {
        const seconds = nanos / BigInt(1e9);
        const remainderNanos = nanos % BigInt(1e9);
        // After normalization, we can safely cast to number.
        return new Duration(seconds, Number(remainderNanos));
    }
}

// The Duration class can also be used as a timestamp.
export { Duration as Timestamp };
