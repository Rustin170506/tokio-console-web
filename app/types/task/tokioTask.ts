import { Duration, Timestamp } from "../common/duration";
import type { Field } from "../common/field";
import { Warning, type Linter, type Warn } from "../warning/warn";
import type { TokioTaskStats } from "./tokioTaskStats";

export enum TaskState {
    Completed,
    Running,
    Scheduled,
    Idle,
}

export enum TaskLintResult {
    Linted,
    RequiresRecheck,
}

export class TokioTask {
    // The task's pretty (console-generated, sequential) task ID.
    //
    // This is NOT the `tracing::span::Id` for the task's tracing span on the
    // remote.
    id: bigint;
    // The `tokio::task::Id` in the remote tokio runtime.
    taskId?: bigint;
    // The `tracing::span::Id` on the remote process for this task's span.
    //
    // This is used when requesting a task details stream
    spanId: bigint;
    // A precomputed short description string used in the async ops table
    shortDesc: string;
    // Fields that don't have their own column, pre-formatted
    fields: Array<Field>;
    // The task statistics that are updated over the lifetime of the task.
    stats: TokioTaskStats;
    // The target of the span representing the task.
    target: string;
    // The name of the task
    name?: string;
    // The source file and line number the task was spawned from
    location: string;
    // The kind of task, currently one of task, blocking, block_on, local
    kind: string;
    // Currently active warnings for this task.
    /* eslint-disable no-use-before-define */
    warnings: Array<Warn<TokioTask>>;

    constructor(
        id: bigint,
        taskId: bigint | undefined,
        spanId: bigint,
        shortDesc: string,
        formattedFields: Array<Field>,
        stats: TokioTaskStats,
        target: string,
        name: string | undefined,
        location: string,
        kind: string,
    ) {
        this.id = id;
        this.taskId = taskId;
        this.spanId = spanId;
        this.shortDesc = shortDesc;
        this.fields = formattedFields;
        this.stats = stats;
        this.target = target;
        this.name = name;
        this.location = location;
        this.kind = kind;
        this.warnings = [];
    }

    totalDuration(since: Timestamp): Duration {
        const duration = since.subtract(this.stats.createdAt);
        return this.stats.total ?? duration;
    }

    busyDuration(since: Timestamp): Duration {
        if (this.stats.lastPollStarted) {
            if (
                !this.stats.lastPollEnded ||
                this.stats.lastPollStarted > this.stats.lastPollEnded
            ) {
                // In this case the task is being polled at the moment.
                const currentTimeInPollDuration = since.subtract(
                    this.stats.lastPollStarted,
                );
                return this.stats.busy.add(currentTimeInPollDuration);
            }
        }
        return this.stats.busy;
    }

    scheduledDuration(since: Timestamp): Duration {
        if (this.stats.lastWake) {
            if (
                !this.stats.lastPollStarted ||
                this.stats.lastWake > this.stats.lastPollStarted
            ) {
                // In this case the task is scheduled, but has not yet been polled.
                const currentTimeSinceWakeDuration = since.subtract(
                    this.stats.lastWake,
                );
                return this.stats.scheduled.add(currentTimeSinceWakeDuration);
            }
        }
        return this.stats.scheduled;
    }

    idleDuration(since: Timestamp): Duration {
        if (this.stats.idle) {
            return this.stats.idle;
        } else {
            const total = this.totalDuration(since);
            const busy = this.busyDuration(since);
            const scheduled = this.scheduledDuration(since);
            if (total.greaterThan(busy.add(scheduled))) {
                return total.subtract(busy.add(scheduled));
            }
        }
        return new Duration(BigInt(0), 0);
    }

    isRunning(): boolean {
        if (this.stats.lastPollStarted && this.stats.lastPollEnded) {
            return this.stats.lastPollStarted.greaterThan(
                this.stats.lastPollEnded,
            );
        }
        if (
            this.stats.lastPollStarted &&
            this.stats.lastPollEnded === undefined
        ) {
            return true;
        }
        return false;
    }

    isScheduled(): boolean {
        if (this.stats.lastWake && this.stats.lastPollStarted) {
            return this.stats.lastWake.greaterThan(this.stats.lastPollStarted);
        }
        return false;
    }

    isCompleted(): boolean {
        return this.stats.total !== undefined;
    }

    isBlocking(): boolean {
        return ["block_on", "blocking"].includes(this.kind);
    }

    // Returns whether this task has signaled via its waker to run again.
    // Once the task has been polled, this is changed back to false.
    isAwakened(): boolean {
        // Before the first poll, the task is waiting on the executor to run it
        // for the first time.
        if (this.totalPolls() === 0n) {
            return true;
        }

        if (this.stats.lastWake && this.stats.lastPollStarted) {
            return this.stats.lastWake.greaterThan(this.stats.lastPollStarted);
        }

        if (this.stats.lastWake && this.stats.lastPollStarted === undefined) {
            return true;
        }

        return false;
    }

    totalPolls(): bigint {
        return this.stats.polls;
    }

    state(): TaskState {
        if (this.isCompleted()) {
            return TaskState.Completed;
        }

        if (this.isRunning()) {
            return TaskState.Running;
        }

        if (this.isScheduled()) {
            return TaskState.Scheduled;
        }

        return TaskState.Idle;
    }

    wakerCount(): bigint {
        return this.stats.wakerClones - this.stats.wakerDrops > 0n
            ? this.stats.wakerClones - this.stats.wakerDrops
            : 0n;
    }

    /**
     * Returns the elapsed time since the task was last woken, relative to
     * given `now` timestamp.
     *
     * Returns `undefined` if the task has never been woken, or if it was last woken
     * more recently than `now` (which *shouldn't* happen as long as `now` is the
     * timestamp of the last stats update...)
     */
    sinceWakeDuration(now: Timestamp): Duration | undefined {
        if (this.stats.lastWake) {
            return now.subtract(this.stats.lastWake);
        }
    }

    durationPercent(now: Timestamp, amt: Duration): number {
        const total = this.totalDuration(now);
        let percent = (amt.asSeconds() / total.asSeconds()) * 100;
        if (percent > 100) {
            percent = 100;
        }

        return percent;
    }

    lint(linters: Array<Linter<TokioTask>>): TaskLintResult {
        this.warnings.length = 0;
        let recheck = false;
        linters.forEach((linter) => {
            const result = linter.check(this);
            switch (result) {
                case Warning.Ok:
                    break;
                case Warning.Recheck:
                    recheck = true;
                    break;
                case Warning.Warn:
                    this.warnings.push(linter);
                    break;
            }
        });
        if (recheck) {
            return TaskLintResult.RequiresRecheck;
        } else {
            return TaskLintResult.Linted;
        }
    }
}
