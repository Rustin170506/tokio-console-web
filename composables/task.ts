import { Duration } from "./duration";
import { Location } from "./../gen/common_pb";
import type { Stats as ProtoTaskStats } from "~/gen/tasks_pb";

export interface TaskStats {
    polls: bigint;
    createdAt: Date;
    droppedAt?: Date;
    busy: Duration;
    scheduled: Duration;
    lastPollStarted?: Date;
    lastPollEnded?: Date;
    idle?: Duration;
    total?: Duration;

    // === waker stats ===
    // Total number of times the task has been woken over its lifetime.
    wakes: bigint;
    // Total number of times the task's waker has been cloned
    wakerClones: bigint;
    // Total number of times the task's waker has been dropped.
    wakerDrops: bigint;
    // The timestamp of when the task was last woken.
    lastWake?: Date;
    // Total number of times the task has woken itself.
    selfWakes: bigint;
}

export function fromProtoTaskStats(stats: ProtoTaskStats) {
    const createdAt = stats.createdAt!.toDate();

    const droppedAt = stats.droppedAt?.toDate();
    const total = droppedAt
        ? new Duration(
              BigInt(
                  Math.floor(
                      (droppedAt.getTime() - createdAt.getTime()) / 1000,
                  ),
              ),
              0,
          )
        : undefined;

    const pollStats = stats.pollStats!;
    const busy = pollStats.busyTime
        ? new Duration(pollStats.busyTime.seconds, pollStats.busyTime.nanos)
        : new Duration(BigInt(0), 0);

    const scheduled = stats.scheduledTime
        ? new Duration(stats.scheduledTime.seconds, stats.scheduledTime.nanos)
        : new Duration(BigInt(0), 0);
    const idle = total ? total.subtract(busy).subtract(scheduled) : undefined;

    return {
        polls: pollStats.polls,
        createdAt,
        droppedAt,
        busy,
        scheduled,
        lastPollStarted: pollStats.lastPollStarted?.toDate(),
        lastPollEnded: pollStats.lastPollEnded?.toDate(),
        idle,
        total,
        wakes: stats.wakes,
        wakerClones: stats.wakerClones,
        wakerDrops: stats.wakerDrops,
        lastWake: stats.lastWake?.toDate(),
        selfWakes: stats.selfWakes,
    };
}

export interface Task {
    // The task's pretty (console-generated, sequential) task ID.
    //
    // This is NOT the `tracing::span::Id` for the task's tracing span on the
    // remote.
    id: bigint;
    // The `tokio::task::Id` in the remote tokio runtime.
    taskId?: bigint;
    // The `tracing::span::Id` on the remote process for this task's span.
    //
    // This is used when requesting a task details stream
    spanId: bigint;
    // A precomputed short description string used in the async ops table
    shortDesc: string;
    // Fields that don't have their own column, pre-formatted
    formattedFields: Array<string>;
    // The task statistics that are updated over the lifetime of the task.
    stats: TaskStats;
    // The target of the span representing the task.
    target: string;
    // The name of the task
    name?: string;
    // The source file and line number the task was spawned from
    location: string;
    // The kind of task, currently one of task, blocking, block_on, local
    kind: string;
}

function truncateRegistryPath(s: string): string {
    const regex = /.*\/cargo(\/registry\/src\/[^/]*\/|\/git\/checkouts\/)/;
    return s.replace(regex, "<cargo>/");
}

export function formatLocation(loc?: Location): string {
    if (loc) {
        if (loc.file) {
            const truncated = truncateRegistryPath(loc.file);
            loc.file = truncated;
        }
        return loc.file ?? "<unknown file>";
    }
    return "<unknown location>";
}

export interface TaskData {
    id: bigint;
    name: string;
    state: string;
    total: string;
    busy: string;
    sched: string;
    idle: string;
    pools: bigint;
    kind: string;
    location: string;
    fields: Array<string>;
}

export function toTaskData(task: Task): TaskData {
    return {
        id: task.id,
        name: task.name ?? "",
        // FIXME: use real state.
        state: "running",
        total: (task.stats.total ?? new Duration(BigInt(0), 0)).toString(),
        busy: task.stats.busy.toString(),
        sched: task.stats.scheduled.toString(),
        idle: (task.stats.idle ?? new Duration(BigInt(0), 0)).toString(),
        pools: task.stats.polls,
        kind: task.kind,
        location: task.location,
        fields: task.formattedFields,
    };
}
