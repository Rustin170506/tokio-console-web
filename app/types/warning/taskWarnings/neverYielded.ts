import { TaskState, TokioTask } from "../../task/tokioTask";
import { Duration, Timestamp } from "../../common/duration";
import { type Warn, Warning } from "../warn";

const DEFAULT_DURATION: Duration = new Duration(1n, 0);

// Warning for if a task has never yielded
export class NeverYielded implements Warn<TokioTask> {
    name: string;
    minDuration: Duration;
    description: string;

    constructor(minDuration: Duration = DEFAULT_DURATION) {
        this.name = "NeverYielded";
        this.minDuration = minDuration;
        this.description = `tasks have never yielded (threshold ${minDuration.toString()})`;
    }

    check(task: TokioTask): Warning {
        // Don't fire warning for tasks that are not async.
        if (task.isBlocking()) {
            return Warning.Ok;
        }
        // Don't fire warning for tasks that are waiting to run.
        if (task.state() !== TaskState.Running) {
            return Warning.Ok;
        }

        if (task.totalPolls() > 1n) {
            return Warning.Ok;
        }

        // Avoid short-lived task false positives
        const busy = task.busyDuration(Timestamp.now());
        if (busy.greaterThanOrEqual(this.minDuration)) {
            return Warning.Warn;
        } else {
            return Warning.Recheck;
        }
    }

    format(task: TokioTask): string {
        const busy = task.busyDuration(Timestamp.now()).toString();
        return `This task has never yielded (${busy})`;
    }

    summary(): string {
        return this.description;
    }
}
