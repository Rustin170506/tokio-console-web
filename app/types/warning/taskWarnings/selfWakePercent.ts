import { Warning, type Warn } from "../warn";
import { TokioTask } from "../../task/tokioTask";

const DEFAULT_PERCENT = 50;

export class SelfWakePercent implements Warn<TokioTask> {
    name: string;
    minPercent: number;
    description: string;

    constructor(minPercent: number = DEFAULT_PERCENT) {
        this.name = "NeverYielded";
        this.minPercent = minPercent;
        this.description = `tasks have woken themselves over ${minPercent}% of the time`;
    }

    check(task: TokioTask): Warning {
        // Don't fire warning for tasks that are not async
        if (task.isBlocking()) {
            return Warning.Ok;
        }
        const selfWakes = task.selfWakePercent();
        if (selfWakes > this.minPercent) {
            return Warning.Warn;
        } else {
            return Warning.Ok;
        }
    }

    format(task: TokioTask): string {
        const selfWakes = task.selfWakePercent();
        return `This task has woken itself for more than ${this.minPercent}% of its total wakeups (${selfWakes}%)`;
    }

    summary(): string {
        return this.description;
    }
}
