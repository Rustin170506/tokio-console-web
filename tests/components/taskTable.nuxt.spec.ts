import { expect, it, describe } from "vitest";
import { mountSuspended, mockNuxtImport } from "@nuxt/test-utils/runtime";
import { createTask } from "./createTask";
import TaskTable from "~/components/TaskTable.vue";
import { TokioTask } from "~/types/task/tokioTask";
import { Timestamp } from "~/types/common/duration";
import { NeverYielded } from "~/types/warning/taskWarnings/neverYielded";
import { LostWaker } from "~/types/warning/taskWarnings/lostWaker";
import { SelfWakePercent } from "~/types/warning/taskWarnings/selfWakePercent";

mockNuxtImport("useTasks", () => {
    return () => {
        const pending = ref<boolean>(false);
        const tasksData = ref<Map<bigint, TokioTask>>(new Map());
        const task = createTask();
        task.warnings.push(new NeverYielded());
        task.warnings.push(new LostWaker());
        task.warnings.push(new SelfWakePercent());
        tasksData.value.set(1n, task);
        const lastUpdatedAt = ref<Timestamp>(new Timestamp(1000n, 0));
        return { pending, tasksData, lastUpdatedAt };
    };
});

describe("TaskTable", () => {
    it("renders correctly", async () => {
        const component = await mountSuspended(TaskTable);
        expect(component.element.querySelector("tbody")).toMatchSnapshot();
    });

    it("displays correct number of columns", async () => {
        const component = await mountSuspended(TaskTable);
        const headers = component.findAll("th");
        expect(headers.length).toBe(11);
    });

    it("displays task data correctly", async () => {
        const component = await mountSuspended(TaskTable);
        const rows = component.findAll("tbody tr");
        expect(rows.length).toBe(1);
    });

    it("displays warning icon for tasks with warnings", async () => {
        const component = await mountSuspended(TaskTable);
        const warningIcon = component.find(
            ".i-heroicons\\:exclamation-triangle",
        );
        expect(warningIcon.exists()).toBe(true);
    });

    it("has total column sorted by default", async () => {
        const component = await mountSuspended(TaskTable);
        const totalHeader = component.find('th[aria-sort="descending"]');
        expect(totalHeader.exists()).toBe(true);
        expect(totalHeader.find("span").text()).toBe("Total");
    });
});
