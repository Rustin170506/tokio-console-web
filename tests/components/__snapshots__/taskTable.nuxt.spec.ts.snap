// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskTable View 1`] = `
<tbody
  class="divide-y divide-gray-200 dark:divide-gray-800"
>
  
  <tr
    class="hover:bg-gray-50 dark:hover:bg-gray-800/50 cursor-pointer"
  >
    <!--v-if-->
    
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      1
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      task1
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      <svg
        aria-hidden="true"
        class="icon"
        data-v-a142084e=""
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <path
          d="M5.25 5.653c0-.856.917-1.398 1.667-.986l11.54 6.347a1.125 1.125 0 0 1 0 1.972l-11.54 6.347a1.125 1.125 0 0 1-1.667-.986z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm text-gray-500 dark:text-gray-300"
    >
      
      <p
        class="text-gray-500 dark:text-gray-300 w-10"
      >
        0ns
      </p>
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm text-green-500 dark:text-green-300"
    >
      
      <p
        class="text-green-500 dark:text-green-300 w-10"
      >
        8m 20s 
      </p>
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm text-green-500 dark:text-green-300"
    >
      
      <p
        class="text-green-500 dark:text-green-300 w-10"
      >
        5m 
      </p>
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm text-green-500 dark:text-green-300"
    >
      
      <p
        class="text-green-500 dark:text-green-300 w-10"
      >
        3m 20s 
      </p>
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      100
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      task
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      app.rs/68:10
      
    </td>
    <td
      class="whitespace-nowrap px-3 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      
      <div
        class="mb-1"
      >
        <p>
          <span
            class="text-blue-600 dark:text-blue-400"
          >
            target
          </span>
          <span
            class="text-gray-500 dark:text-gray-400"
          >
            =
          </span>
          <span
            class="text-green-600 dark:text-green-400"
          >
            tokio:task
          </span>
        </p>
      </div>
      
      
    </td>
    
  </tr>
  
</tbody>
`;
